tmpl/wiki/weekly/2015-01-14.md

```diff
@@ -0,0 +1,149 @@
+rebase it and hopes that won't be too much of a problem for anyone.  In
+general we shouldn't do that but this one would take a while to rebase and
+would prefer to merge quickly.  The functor inline patch from 4.02.1 will have
+big performance boost but we should clearly state that we're not dropping
+support for 4.01 though.
+
+In addition, we should be careful to maintain support for OPAM 1.1 or at the
+very least, warn the user if something won't work with v1.1. Many of the
+upstream distros are still on this version and silent failures result in
+confusion and frustration. There are also an increasing number of libraries
+going for 4.02.1 support only, but it would be better to try and maintain
+support for 4.01 as long as possible (for the same reasons as above).
+
+[115]: https://github.com/mirage/mirage-platform/pull/115
+
+
+#### TLS on Xen ####
+
+ThomasL started an [email thread][tls-thread] about getting the TLS stack
+working on Xen.  He's submitted a number of patches for review to different
+libraries and the changes needed are:
+
+1. Add generic error handling for FLOWs, so we can propagate errors reliably.
+2. Fix the page alignment requirements for Netif.
+3. Add TLS support to conduit.
+
+TLS now works if you apply lots of patches and works pretty well (half the
+speed of normal TLS?). Issue now is how to merge without breaking things as
+certain module signatures change.  There was some discussion of functorising
+and the knock-on effect this would have but overall it seems like we can keep
+this as an interface-level problem.  
+
+There's still the question of getting entropy in Xen. We need several ways to
+do this and [xentropyd][] is the one way we currently have if dom0 cooperates.
+This is fine for our infrastructure as we control dom0 (on cubieboards and
+x86).  However, we also need to be able to gather entropy in other ways for
+deployments to third-party infrastructure. The actual code for getting entropy
+is straightforward but how it is then composed needs to be considered.
+
+In general, there's reluctance to release the TLS stuff until there's an
+entropy story so DavidK may take a look at the options.  In the meantime,
+we're happy to have an intermediate patch that *only* uses xentropyd and we
+can then use it for websites hosted on our own infrastructure.
+
+[tls-thread]: http://lists.xenproject.org/archives/html/mirageos-devel/2015-01/msg00066.html
+[mirage-dev]: https://github.com/mirage/mirage-dev
+[xentropyd]: https://github.com/mirage/xentropyd
+
+
+#### Error handling in Mirage ####
+
+ThomasL has been working on this. To make conduit `FLOW` wrap the TLS flows we
+need a way to propagate the errors sensibly, so has added a type to `FLOW`
+that can do this. It might be nice to have an algebraic data type to describe
+an error type, or if we wait a few months until we switch to 4.02.1, then we
+can just use [Open Types][opentypes].  The nice thing with current patch is
```

Comment:
We already have algebraic data types for errors, which is good if you want to match on a particular error from a particular concrete implementation (e.g. `FAT_index_corrupted` or something). In theory, this is slighly more correct than using exceptions because error returns don't propagate automatically. e.g. if you get `Refused` from TCP then you know it was the TCP connection itself that was refused, not some other refusal from another function TCP called that happened to throw the same exception. In practice, this doesn't seem to be a problem except when using very generic exceptions like `Not_found`.

I guess using open types means we can list some common errors in the interfaces but have implementations provide more, which may be useful but doesn't solve my problem: when you *don't* want to handle specific exceptions we need a way to get a human-readable string for display (and this is the common case).

So:
* Replacing error codes with exceptions (which I proposed originally) would work most of the time and would simplify the code greatly, although it's slightly less strict than the current system. On the other hand, the current system encourages people to throw away errors (look at any `main.ml` for examples!) so I'm sure it would be an overall improvement. It would require API changes everywhere, however.
* Adding an `error_message` function (which my current patches do) means that people can still handle errors as before when needed, but can also always get a human-readable string too, so there's never any excuse for throwing away diagnostic information. You do still need to write error handling code everywhere when you just want to propagate exceptions, though, which clutters the code, and old code will continue to discard useful information for some time to come.